services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/controller/SubnetController.java

```diff
@@ -86,59 +101,106 @@ public SubnetStateJson createSubnetState(@PathVariable String projectid, @Reques
+                    throw new CompletionException(e);
+                }
+            }, ThreadPoolExecutorUtils.SELECT_POOL_EXECUTOR);
 
             //Prepare Route Rule(IPv4/6) for Subnet
-            RouteWebJson routeResponse = this.subnetService.createRouteRules(inSubnetState.getId(), inSubnetState);
-            if (routeResponse == null) {
-                throw new FallbackException("fallback request");
-            }
+            CompletableFuture<RouteWebJson> routeFuture = CompletableFuture.supplyAsync(() -> {
+                try {
+                    return this.subnetService.createRouteRules(inSubnetState.getId(), inSubnetState);
+                } catch (Exception e) {
+                    throw new CompletionException(e);
+                }
+            }, ThreadPoolExecutorUtils.SELECT_POOL_EXECUTOR).handle((s, e) -> {
+                routeResponseAtomic.set(s);
+                return s;
+            });;
+
 
-            //Allocate Gateway Mac
-//            MacStateJson macResponse = this.subnetService.allocateMacGateway(projectid, inSubnetState.getVpcId(), portId);
-//            if (macResponse == null) {
-//                throw new ResourcePersistenceException();
-//            }
 
             // Verify/Allocate Gateway IP, subnet id, port id, subnet cidr, response:IP - unique
 //            IPStateJson ipResponse = this.subnetService.allocateIPGateway(inSubnetState.getId(), inSubnetState.getCidr(), portId);
 //            if (ipResponse == null) {
 //                throw new ResourcePersistenceException();
 //            }
 
+            // Synchronous blocking
+            CompletableFuture<Void> allFuture = CompletableFuture.allOf(vpcFuture, macFuture, routeFuture);
+            allFuture.join();
+
+            macResponse = macFuture.join();
+            routeResponse = routeFuture.join();
+            logger.info("Total processing time:" + (System.currentTimeMillis() - start) + "ms");
+
             // set up value of properties for subnetState
-            List<RouteWebObject> routes = subnetState.getRoutes();
+            List<RouteWebObject> routes = inSubnetState.getRoutes();
             if (routes == null) {
                 routes = new ArrayList<>();
             }
             routes.add(routeResponse.getRoute());
-            subnetState.setRoutes(routes);
+            inSubnetState.setRoutes(routes);
             //subnetState.setGatewayIp(ipResponse.getIpState().getIp());
 
+            this.subnetDatabaseService.addSubnet(inSubnetState);
+
+            subnetState = this.subnetDatabaseService.getBySubnetId(inSubnetState.getId());
+            if (subnetState == null) {
+                throw new ResourcePersistenceException();
+            }
+
+            return new SubnetStateJson(subnetState);
+
         } catch (ResourcePersistenceException e) {
+            logger.error(e.getMessage());
             throw new Exception(e);
-        } catch (FallbackException e) {
```

Comment:
We also need to handle FallbackException when the downstream services or db fall back fail. 
In that case, we need to log error or critical as well.